### ISSUE TITLE
key_parsing_mask_list_align why only one face part is being used? 

### ISSUE THREAD

**Issue by trofimovaolga** [JackAILab/ConsistentID#30]
Hi, I'm trying to understand the code and I'd really appreciate if someone could help me understand how face parts are used.
You are using 5 parts: face, eyes, ears, nose and lips (mouth). 
Inside of get_prepare_facemask method you are obtaining key_parsing_mask_list - a list with mask-images of each face part, but there is only one eye out of two (right of left), one lip out of two, one ear out of two. Could you please help me understand why?
Further in code these parts are used as inputs for CLIPEncoder and I wonder why not all parts are used.

**Comment by JackAILab** [comment]
This is because most of the time, llava does not distinguish between these paired facial features when describing facial features information (for example, the descriptions of left and right ears are the same). Therefore, we merge the paired facial features descriptions together and retain only one mask among the paired facial features.

In addition, due to the large amount of training data, not all facial features can be obtained. If all facial features components are regarded as input to CLIPEncoder, there will be a large number of redundant None values. Therefore, we only send the retained facial features images to CLIPEncoder.

Of course, you can also try to inject more facial features into the model, and perhaps the ID consistency can be improved. If you have any further ideas, please feel free to ask questions and PR.

**Comment by trofimovaolga** [comment]
I tried to ignore this whole story with segmentation during inference and only used text encoding features without FacialEncoder. The results are pretty good, it is hard to say if it is better or worse, I'd say it is the same.
![image](https://github.com/JackAILab/ConsistentID/assets/11677412/634f2ab6-d673-48c3-a8d2-5a7398de7670)

**Comment by Jack-Crum** [comment]
> I tried to ignore this whole story with segmentation during inference and only used text encoding features without FacialEncoder. The results are pretty good, it is hard to say if it is better or worse, I'd say it is the same. ![image](https://private-user-images.githubusercontent.com/11677412/330832867-634f2ab6-d673-48c3-a8d2-5a7398de7670.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzNjQ5NjcsIm5iZiI6MTcxNjM2NDY2NywicGF0aCI6Ii8xMTY3NzQxMi8zMzA4MzI4NjctNjM0ZjJhYjYtZDY3My00OGMzLWE4ZDItNWE3Mzk4ZGU3NjcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIyVDA3NTc0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwYzNhMTQ5MTMxZmM0Yzc5YzlkMDE4ZDg0ZThjMjJlMDljMzA1NDhlNmUzYTA2YzNjZWEyYWJhZDM4NTI4YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9nj8PcHO5WogoY04XL9ZgKWXNZpjvJtDP2aN0d-oikk)

Your observation is precise. In fact, both FaceID and FacialEncoder can maintain ID consistency, and their simultaneous presence ensures more stable ID consistency. For more details, you can refer to the ablation experiments in the paper.

Specifically, our model introduces two types of ID consistency information. One type is FaceID, which is used to maintain the global consistency of the ID structure; however, the extraction of FaceID can be unstable. The other type is FacialEncoder, which enhances the consistency of facial features. According to experimental results on a large-scale dataset, FacialEncoder contributes more consistently to ID maintenance. This means that even when FaceID is removed, the model can still achieve more stable ID consistency with the help of FacialEncoder.

If you have any further thoughts, feel free to share and submit a PR~

**Comment by JackAILab** [comment]
Your observation is precise. In fact, both FaceID and FacialEncoder can maintain ID consistency, and their simultaneous presence ensures more stable ID consistency. For more details, you can refer to the ablation experiments in the paper.

Specifically, our model introduces two types of ID consistency information. One type is FaceID, which is used to maintain the global consistency of the ID structure; however, the extraction of FaceID can be unstable. The other type is FacialEncoder, which enhances the consistency of facial features. According to experimental results on a large-scale dataset, FacialEncoder contributes more consistently to ID maintenance. This means that even when FaceID is removed, the model can still achieve more stable ID consistency with the help of FacialEncoder.

If you have any further thoughts, feel free to share and submit a PR.
